File: services/user-service/src/main/java/com/visor/school/userservice/controller/PermissionController.java
package com.visor.school.userservice.controller;

import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.visor.school.common.api.ApiResponse;
import com.visor.school.userservice.dto.CreatePermissionRequest;
import com.visor.school.userservice.dto.PermissionResponse;
import com.visor.school.userservice.model.Permission;
import com.visor.school.userservice.service.PermissionService;

import jakarta.validation.Valid;

/**
 * Permission management controller
 * Requires ADMINISTRATOR role for permission management operations
 */
@RestController
@RequestMapping("/v1/permissions")
public class PermissionController {

    private final PermissionService permissionService;

    public PermissionController(PermissionService permissionService) {
        this.permissionService = permissionService;
    }

    /**
     * Create a new permission
     */
    @PostMapping
    @PreAuthorize("hasRole('ADMINISTRATOR')")
    public ResponseEntity<ApiResponse<PermissionResponse>> createPermission(@Valid @RequestBody CreatePermissionRequest request) {
        Permission permission = permissionService.createPermission(
            request.permissionKey(),
            request.name(),
            request.description(),
            request.category()
        );

        return ResponseEntity.status(HttpStatus.CREATED)
            .body(ApiResponse.success(PermissionResponse.from(permission), "Permission created successfully"));
    }

    /**
     * Get all permissions
     */
    @GetMapping
    @PreAuthorize("hasRole('ADMINISTRATOR')")
    public ResponseEntity<ApiResponse<List<PermissionResponse>>> getAllPermissions() {
        List<Permission> permissions = permissionService.getAllPermissions();
        return ResponseEntity.ok(ApiResponse.success(permissions.stream().map(PermissionResponse::from).collect(Collectors.toList())));
    }

    /**
     * Get permission by key
     */
    @GetMapping("/{permissionKey}")
    @PreAuthorize("hasRole('ADMINISTRATOR')")
    public ResponseEntity<ApiResponse<PermissionResponse>> getPermission(@PathVariable String permissionKey) {
        Permission permission = permissionService.findByKey(permissionKey);
        if (permission == null) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.ok(ApiResponse.success(PermissionResponse.from(permission)));
    }

    /**
     * Get permissions by category
     */
    @GetMapping("/category/{category}")
    @PreAuthorize("hasRole('ADMINISTRATOR')")
    public ResponseEntity<ApiResponse<List<PermissionResponse>>> getPermissionsByCategory(@PathVariable String category) {
        List<Permission> permissions = permissionService.getByCategory(category);
        return ResponseEntity.ok(ApiResponse.success(permissions.stream().map(PermissionResponse::from).collect(Collectors.toList())));
    }

    /**
     * Assign permission to user
     */
    @PostMapping("/{permissionId}/assign/{userId}")
    @PreAuthorize("hasRole('ADMINISTRATOR')")
    public ResponseEntity<ApiResponse<Map<String, String>>> assignPermissionToUser(
        @PathVariable UUID permissionId,
        @PathVariable UUID userId
    ) {
        // assignedBy logic omitted in Kotlin controller, likely not passed or handled by service if overloaded? 
        // Service: assignToUser(userId, permissionId, assignedBy)
        // Kotlin controller: permissionService.assignToUser(userId, permissionId)
        // Kotlin service parameter: assignedBy: UUID? = null
        // My Java service matches: assignToUser(userId, permissionId, assignedBy)
        // I'll pass null for assignedBy as in Kotlin.
        
        permissionService.assignToUser(userId, permissionId, null);

        return ResponseEntity.ok(
            ApiResponse.success(
                Map.of(
                    "message", "Permission assigned successfully",
                    "userId", userId.toString(),
                    "permissionId", permissionId.toString()
                )
            )
        );
    }

    /**
     * Get permissions for a user
     */
    @GetMapping("/user/{userId}")
    @PreAuthorize("hasRole('ADMINISTRATOR') or @securityContextService.isCurrentUserId(#userId)")
    public ResponseEntity<ApiResponse<List<PermissionResponse>>> getUserPermissions(@PathVariable UUID userId) {
        List<Permission> permissions = permissionService.getByUser(userId);
        return ResponseEntity.ok(ApiResponse.success(permissions.stream().map(PermissionResponse::from).collect(Collectors.toList())));
    }

    /**
     * Remove permission from user
     */
    @DeleteMapping("/{permissionId}/user/{userId}")
    @PreAuthorize("hasRole('ADMINISTRATOR')")
    public ResponseEntity<ApiResponse<Map<String, String>>> removePermissionFromUser(
        @PathVariable UUID permissionId,
        @PathVariable UUID userId
    ) {
        permissionService.removeFromUser(userId, permissionId);

        return ResponseEntity.ok(
            ApiResponse.success(
                Map.of("message", "Permission removed successfully")
            )
        );
    }
}

File: services/user-service/src/main/kotlin/com/visor/school/userservice/controller/PermissionController.kt
package com.visor.school.userservice.controller

import com.visor.school.common.api.ApiResponse
import com.visor.school.userservice.model.Permission
import com.visor.school.userservice.service.PermissionService
import jakarta.validation.Valid
import jakarta.validation.constraints.NotBlank
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*
import java.util.UUID

/**
 * Permission management controller
 * Requires ADMINISTRATOR role for permission management operations
 */
@RestController
@RequestMapping("/api/v1/permissions")
class PermissionController(
    private val permissionService: PermissionService
) {

    /**
     * Create a new permission
     */
    @PostMapping
    @PreAuthorize("hasRole('ADMINISTRATOR')")
    fun createPermission(@Valid @RequestBody request: CreatePermissionRequest): ResponseEntity<ApiResponse<PermissionResponse>> {
        val permission = permissionService.createPermission(
            permissionKey = request.permissionKey,
            name = request.name,
            description = request.description,
            category = request.category
        )

        return ResponseEntity.status(HttpStatus.CREATED)
            .body(ApiResponse.success(PermissionResponse.from(permission), "Permission created successfully"))
    }

    /**
     * Get all permissions
     */
    @GetMapping
    @PreAuthorize("hasRole('ADMINISTRATOR')")
    fun getAllPermissions(): ResponseEntity<ApiResponse<List<PermissionResponse>>> {
        val permissions = permissionService.getAllPermissions()
        return ResponseEntity.ok(ApiResponse.success(permissions.map { PermissionResponse.from(it) }))
    }

    /**
     * Get permission by key
     */
    @GetMapping("/{permissionKey}")
    @PreAuthorize("hasRole('ADMINISTRATOR')")
    fun getPermission(@PathVariable permissionKey: String): ResponseEntity<ApiResponse<PermissionResponse>> {
        val permission = permissionService.findByKey(permissionKey)
            ?: return ResponseEntity.notFound().build()

        return ResponseEntity.ok(ApiResponse.success(PermissionResponse.from(permission)))
    }

    /**
     * Get permissions by category
     */
    @GetMapping("/category/{category}")
    @PreAuthorize("hasRole('ADMINISTRATOR')")
    fun getPermissionsByCategory(@PathVariable category: String): ResponseEntity<ApiResponse<List<PermissionResponse>>> {
        val permissions = permissionService.getByCategory(category)
        return ResponseEntity.ok(ApiResponse.success(permissions.map { PermissionResponse.from(it) }))
    }

    /**
     * Assign permission to user
     */
    @PostMapping("/{permissionId}/assign/{userId}")
    @PreAuthorize("hasRole('ADMINISTRATOR')")
    fun assignPermissionToUser(
        @PathVariable permissionId: UUID,
        @PathVariable userId: UUID
    ): ResponseEntity<ApiResponse<Map<String, String>>> {
        permissionService.assignToUser(userId, permissionId)

        return ResponseEntity.ok(
            ApiResponse.success(
                mapOf(
                    "message" to "Permission assigned successfully",
                    "userId" to userId.toString(),
                    "permissionId" to permissionId.toString()
                )
            )
        )
    }

    /**
     * Get permissions for a user
     */
    @GetMapping("/user/{userId}")
    @PreAuthorize("hasRole('ADMINISTRATOR') or #userId == authentication.principal.id")
    fun getUserPermissions(@PathVariable userId: UUID): ResponseEntity<ApiResponse<List<PermissionResponse>>> {
        val permissions = permissionService.getByUser(userId)
        return ResponseEntity.ok(ApiResponse.success(permissions.map { PermissionResponse.from(it) }))
    }

    /**
     * Remove permission from user
     */
    @DeleteMapping("/{permissionId}/user/{userId}")
    @PreAuthorize("hasRole('ADMINISTRATOR')")
    fun removePermissionFromUser(
        @PathVariable permissionId: UUID,
        @PathVariable userId: UUID
    ): ResponseEntity<ApiResponse<Map<String, String>>> {
        permissionService.removeFromUser(userId, permissionId)

        return ResponseEntity.ok(
            ApiResponse.success(
                mapOf("message" to "Permission removed successfully")
            )
        )
    }
}

// Request DTOs
data class CreatePermissionRequest(
    @field:NotBlank
    val permissionKey: String,

    @field:NotBlank
    val name: String,

    @field:NotBlank
    val description: String,

    @field:NotBlank
    val category: String
)

// Response DTOs
data class PermissionResponse(
    val id: UUID,
    val permissionKey: String,
    val name: String,
    val description: String,
    val category: String
) {
    companion object {
        fun from(permission: Permission): PermissionResponse {
            return PermissionResponse(
                id = permission.id,
                permissionKey = permission.permissionKey,
                name = permission.name,
                description = permission.description,
                category = permission.category
            )
        }
    }
}

